import struct
from datetime import datetime
from time import sleep

from PySide6.QtWidgets import (
    QMainWindow, QSystemTrayIcon, QMenu, QPushButton, QMessageBox,
    QLabel, QTextEdit, QApplication
)
from PySide6.QtGui import QIcon, QAction, QPixmap, QPainter, QFont, QFontMetrics
from PySide6.QtCore import Qt, QTimer, QFile, QThread
from PySide6.QtUiTools import QUiLoader
from PySide6.QtWidgets import QFrame

# --- æœ¬åœ°æ¨¡å—å¯¼å…¥ ---
from serial_worker import SerialWorker
from config_handler import ConfigHandler
from database_handler import DatabaseHandler
from history_window import HistoryWindow
import constants as const
from mouse_handler import MouseDataProcessor
from utils import resource_path 


LOGGING_ENABLED = True


class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        
        # 1. è¿è¡Œæ—¶åŠ è½½ UI
        self._load_ui(resource_path("main.ui")) # <--- ä½¿ç”¨æ–°å‡½æ•°
        self.setFixedSize(self.size())          # ç¦æ­¢è°ƒæ•´çª—å£å¤§å°
        self._center_window()                   # çª—å£å±…ä¸­
        self.setAttribute(Qt.WidgetAttribute.WA_QuitOnClose, True)
        self.setWindowFlags(Qt.WindowMinimizeButtonHint | Qt.WindowCloseButtonHint)

        # --- å›¾æ ‡ ---
        self.icon_heart = self._create_emoji_icon('â¤ï¸')
        self.icon_white_heart = self._create_emoji_icon('ğŸ©¶')
        self.is_heart_icon = False
        self.setWindowIcon(self.icon_heart)
        
        # --- ç»‘å®š UI æ§ä»¶ ---
        self.start_button = self.findChild(QPushButton, "btn_start")
        if self.start_button:
            self.start_button.clicked.connect(self.on_start_button_clicked)
            
        # --- ç»‘å®šå†å²æ•°æ®æŒ‰é’® ---
        self.history_button = self.findChild(QPushButton, "btn_history")
        if self.history_button:
            self.history_button.clicked.connect(self.show_history_window)
        else:
            print("è­¦å‘Š: æœªåœ¨ UI æ–‡ä»¶ä¸­æ‰¾åˆ°åä¸º 'btn_history' çš„ QPushButtonã€‚")
        
        # --- ç»‘å®šåˆ·æ–°é¼ æ ‡æ•°æ®æŒ‰é’® ---
        self.mousedata_button = self.findChild(QPushButton, "btn_mousedata")
        if self.mousedata_button:
            self.mousedata_button.clicked.connect(self.on_mousedata_button_clicked)
        else:
            print("è­¦å‘Š: æœªåœ¨ UI æ–‡ä»¶ä¸­æ‰¾åˆ°åä¸º 'btn_mousedata' çš„ QPushButtonã€‚")
            
        self.metric_keys = [
            'heartrate', 'spo2', 'bk', 'fatigue', 'systolic', 'diastolic', 
            'cardiac', 'resistance', 'rr_interval', 'sdnn', 'rmssd', 
            'nn50', 'pnn50', 'timestamp'
        ]
        
        ui_label_names = {
            'heartrate': 'label_hr_value', 'spo2': 'label_spo2_value',
            'bk': 'label_mc_value', 'fatigue': 'label_fi_value',
            'systolic': 'label_sbp_value', 'diastolic': 'label_dbp_value',
            'cardiac': 'label_co_value', 'resistance': 'label_pr_value'
        }

        self.value_labels = {}
        for key, name in ui_label_names.items():
            label = self.findChild(QLabel, name)
            if label:
                self.value_labels[key] = label
                if key in const.HEALTH_METRICS_TOOLTIPS:
                    label.setToolTip(const.HEALTH_METRICS_TOOLTIPS[key])
            else:
                print(f"è­¦å‘Š: æœªåœ¨ UI æ–‡ä»¶ä¸­æ‰¾åˆ°åä¸º '{name}' çš„ QLabelã€‚")

        # é¼ æ ‡ç»Ÿè®¡å€¼æ ‡ç­¾
        self.label_distance = self.findChild(QLabel, "label_distance")
        self.label_leftclick = self.findChild(QLabel, "label_leftclick")
        self.label_midclick = self.findChild(QLabel, "label_midclick")
        self.label_rightclick = self.findChild(QLabel, "label_rightclick")

        self.log_output = self.findChild(QTextEdit, "log_output")
        if self.log_output:
            self.log_output.setReadOnly(True)

        # --- ç³»ç»Ÿæ‰˜ç›˜ ---
        self.tray_icon = QSystemTrayIcon(self)
        self.tray_icon.setIcon(self.icon_heart)
        tray_menu = QMenu()
        show_action = QAction("æ˜¾ç¤ºä¸»ç•Œé¢", self)
        about_action = QAction("å…³äº", self)
        exit_action = QAction("é€€å‡º", self)
        
        tray_menu.addAction(show_action)
        tray_menu.addAction(about_action)
        tray_menu.addSeparator()
        tray_menu.addAction(exit_action)
        
        self.tray_icon.setContextMenu(tray_menu)
        self.tray_icon.activated.connect(self._handle_tray_activation)
        show_action.triggered.connect(self.show_window)
        about_action.triggered.connect(self._show_about_dialog)
        exit_action.triggered.connect(self.exit_app)
        self.tray_icon.show()

        # --- å®šæ—¶å™¨ ---
        self.blink_timer = QTimer(self)
        self.blink_timer.setInterval(500)
        self.blink_timer.timeout.connect(self._toggle_icon)
        
        self.detection_timeout_timer = QTimer(self)
        self.detection_timeout_timer.setSingleShot(True)
        self.detection_timeout_timer.setInterval(100 * 1000)  # å»¶é•¿è¶…æ—¶åˆ°100ç§’
        self.detection_timeout_timer.timeout.connect(self.on_detection_timeout)
        
        # å¼€å§‹ä½“æ£€å€’è®¡æ—¶ï¼ˆACK åå¯åŠ¨ï¼‰
        self.countdown_timer = QTimer(self)
        self.countdown_timer.setInterval(1000)
        self.countdown_timer.timeout.connect(self._on_countdown_tick)
        self.countdown_remaining = 0
        
        # --- çŠ¶æ€æ  ---
        self._init_status_bar()

        # --- ä¸šåŠ¡é€»è¾‘å¤„ç†å™¨ ---
        self.config_handler = ConfigHandler()
        self.db_handler = DatabaseHandler(metric_keys=self.metric_keys)
        self._init_serial()

        self.startup_data_loaded = False
        self.history_window_instance = None # ç”¨äºæŒæœ‰å†å²çª—å£çš„å®ä¾‹
        self.startup_sequence()
        
        # é¼ æ ‡æ•°æ®å¤„ç†å™¨
        self.mouse_processor = MouseDataProcessor(self.db_handler)

    def _init_serial(self):
        """åˆå§‹åŒ–ä¸²å£å·¥ä½œçº¿ç¨‹"""
        self.serial_thread = QThread()
        self.serial_worker = SerialWorker()
        self.serial_worker.moveToThread(self.serial_thread)

        self.serial_worker.error_occurred.connect(self._show_error)
        self.serial_worker.log_message.connect(self._log_to_ui)
        self.serial_worker.ack_received.connect(self.on_ack_received)
        self.serial_worker.health_data_received.connect(self.on_health_data_received)
        self.serial_worker.mouse_data_received.connect(self.on_mouse_data_received)
        self.serial_worker.connected.connect(self._update_status_connected)
        self.serial_worker.disconnected.connect(self._update_status_disconnected)

        self.serial_thread.started.connect(self.serial_worker.run)
        # self.serial_worker.disconnected.connect(self.serial_thread.quit) # åœ¨è‡ªåŠ¨é‡è¿æ¨¡å¼ä¸‹ï¼Œçº¿ç¨‹ä¸åº”è½»æ˜“é€€å‡º
        
    def startup_sequence(self):
        """åº”ç”¨å¯åŠ¨æ—¶çš„æ“ä½œåºåˆ—"""
        self._log_to_ui("åº”ç”¨å¯åŠ¨... ä¼˜å…ˆä»è®¾å¤‡è·å–æœ€æ–°æ•°æ®ã€‚")
        self.startup_data_loaded = False
        self.startup_mouse_loaded = False
        
        try:
            com_port = self.config_handler.get_com_port()
            self.serial_worker.connect_serial(com_port)
            
            if self.serial_worker.is_running and not self.serial_thread.isRunning():
                self.serial_thread.start()

            
            QTimer.singleShot(100, lambda: self.serial_worker.send_frame(const.CMD_GET_LAST_HEALTH_DATA))
            QTimer.singleShot(120, lambda: self.serial_worker.send_frame(const.CMD_GET_MOUSE_DATA))
            QTimer.singleShot(5000, self.check_startup_data)

        except Exception as e:
            self._log_to_ui(f"å¯åŠ¨æ—¶è¿æ¥ä¸²å£å¤±è´¥: {e}ã€‚å°è¯•ä»æœ¬åœ°æ–‡ä»¶åŠ è½½...")
            self._load_history_from_db()
            self._load_mouse_from_db()

    def check_startup_data(self):
        """åœ¨å¯åŠ¨è¶…æ—¶åæ£€æŸ¥æ•°æ®æ˜¯å¦å·²åŠ è½½"""
        if not self.startup_data_loaded:
            self._log_to_ui("ä»è®¾å¤‡è·å–æ•°æ®è¶…æ—¶ï¼Œå°è¯•ä»æœ¬åœ°æ–‡ä»¶åŠ è½½...")
            self._load_history_from_db()
        if not self.startup_mouse_loaded:
            self._log_to_ui("ä»è®¾å¤‡è·å–é¼ æ ‡æ•°æ®è¶…æ—¶ï¼Œå°è¯•ä»æ•°æ®åº“åŠ è½½...")
            self._load_mouse_from_db()

    def _load_history_from_db(self):
        """ä»æ•°æ®åº“è¯»å–å¹¶æ˜¾ç¤ºæœ€åä¸€æ¡å†å²æ•°æ®"""
        last_record = self.db_handler.load_last_record()
        if last_record:
            timestamp = last_record.pop('created_at')
            self._log_to_ui(f"ä»æ•°æ®åº“åŠ è½½å†å²æ•°æ® ({timestamp}): {last_record}")
            self._update_data_labels(last_record)
            self.startup_data_loaded = True
        else:
            self._log_to_ui("æ•°æ®åº“ä¸­æ— å†å²æ•°æ®ã€‚")

    def _load_mouse_from_db(self):
        """ä»æ•°æ®åº“è¯»å–å¹¶æ˜¾ç¤ºé¼ æ ‡ç´¯è®¡æ•°æ®"""
        mouse = self.db_handler.load_mouse_data()
        if mouse:
            # ä½¿ç”¨å¤„ç†å™¨è¿›è¡Œå•ä½è½¬æ¢è¾“å‡º
            distance_m = None
            try:
                distance_m = self.mouse_processor.pixels_to_meters_str(mouse['distance'])
            except Exception:
                distance_m = None
            if distance_m:
                self._log_to_ui(
                    f"ä»æ•°æ®åº“åŠ è½½é¼ æ ‡æ•°æ® ({mouse['created_at']}): è·ç¦»={mouse['distance']}px (~{distance_m}), "
                    f"L={mouse['left_click']}, M={mouse['mid_click']}, R={mouse['right_click']}"
                )
            else:
                self._log_to_ui(
                    f"ä»æ•°æ®åº“åŠ è½½é¼ æ ‡æ•°æ® ({mouse['created_at']}): è·ç¦»={mouse['distance']}px, "
                    f"L={mouse['left_click']}, M={mouse['mid_click']}, R={mouse['right_click']}"
                )
            self._update_mouse_labels(mouse['distance'], mouse['left_click'], mouse['mid_click'], mouse['right_click'])
            self.startup_mouse_loaded = True
        else:
            self._log_to_ui("æ•°æ®åº“ä¸­æ— é¼ æ ‡æ•°æ®ã€‚")

    def on_start_button_clicked(self):
        """å¤„ç†å¼€å§‹æŒ‰é’®ç‚¹å‡»äº‹ä»¶"""
        self._log_to_ui("ç‚¹å‡»äº†å¼€å§‹æŒ‰é’®...")
        self._start_blinking()
        self.detection_timeout_timer.start()
        if self.start_button:
            self.start_button.setEnabled(False)
            self.start_button.setText("ä½“æ£€ä¸­...")
        try:
            com_port = self.config_handler.get_com_port()
            if not self.serial_worker.serial_port or not self.serial_worker.serial_port.is_open:
                self.serial_worker.connect_serial(com_port)
                if self.serial_worker.is_running and not self.serial_thread.isRunning():
                    self.serial_thread.start()
                QTimer.singleShot(100, lambda: self.serial_worker.send_frame(const.CMD_START_HEALTH_CHECK))
            else:
                self.serial_worker.send_frame(const.CMD_START_HEALTH_CHECK)
        except Exception as e:
            self._show_error(f"å¼€å§‹ä½“æ£€å¤±è´¥: {e}")
            if self.start_button:
                self.start_button.setEnabled(True)
                self.start_button.setText("å¼€å§‹ä½“æ£€")

    def on_mousedata_button_clicked(self):
        """å¤„ç†åˆ·æ–°é¼ æ ‡æ•°æ®æŒ‰é’®ç‚¹å‡»äº‹ä»¶"""
        self._log_to_ui("åˆ·æ–°é¼ æ ‡æ•°æ®...")
        try:
            com_port = self.config_handler.get_com_port()
            if not self.serial_worker.serial_port or not self.serial_worker.serial_port.is_open:
                self.serial_worker.connect_serial(com_port)
                if self.serial_worker.is_running and not self.serial_thread.isRunning():
                    self.serial_thread.start()
            self.serial_worker.send_frame(const.CMD_GET_MOUSE_DATA)
        except Exception as e:
            self._show_error(f"åˆ·æ–°é¼ æ ‡æ•°æ®å¤±è´¥: {e}")

    def _show_error(self, message: str):
        # QMessageBox.critical(self, "é”™è¯¯", message) # ç¦ç”¨å¼¹çª—
        self._log_to_ui(f"é”™è¯¯: {message}")
        # å‘ç”Ÿä»»ä½•ä¸¥é‡é”™è¯¯æ—¶ï¼Œéƒ½åº”å°è¯•é‡ç½®çŠ¶æ€
        self._reset_detection_state()

    def _show_about_dialog(self):
        """æ˜¾ç¤ºå…³äºå¯¹è¯æ¡†"""
        QMessageBox.about(self, 
            "å…³äº", 
            "<p style='font-size: 1px;'>&nbsp;</p>"
            "<p style='font-size: 14px; font-weight: bold;'> CyMouse æ•°æ®æŸ¥çœ‹å·¥å…· v1.0 &nbsp;</p>"
            "<p align='center'>Powered by <a href='https://cynix.cc' style='color: #89b4fa;'>Cynix.cc</a>&nbsp;&nbsp;&nbsp;</p>"
        )

    def _log_to_ui(self, message: str):
        if not LOGGING_ENABLED:
            return
            
        if self.log_output:
            self.log_output.append(f"[{datetime.now().strftime('%H:%M:%S')}] {message}")
        else:
            print(message)

    def on_ack_received(self, original_cmd: int, status_code: int):
        self._log_to_ui(f"æ”¶åˆ° ACK: åŸå§‹å‘½ä»¤={hex(original_cmd)}, çŠ¶æ€ç ={status_code}")
        if original_cmd == const.CMD_START_HEALTH_CHECK:
            if status_code == const.ACK_SUCCESS:
                self._log_to_ui("è®¾å¤‡å·²ç¡®è®¤å¼€å§‹å¥åº·ç›‘æµ‹ã€‚ç­‰å¾…æ•°æ®...")
                # å¯åŠ¨ 90 ç§’å€’è®¡æ—¶
                self.countdown_remaining = 90
                if self.start_button:
                    self.start_button.setEnabled(False)
                    self.start_button.setText(f"{self.countdown_remaining}ç§’")
                if not self.countdown_timer.isActive():
                    self.countdown_timer.start()
            elif status_code == const.ACK_DEVICE_BUSY:
                self._log_to_ui("è®¾å¤‡æ­£å¿™ï¼Œè¯·ç¨åå†è¯•ã€‚")
                self._reset_detection_state()
            elif status_code == const.ACK_UNKNOWN_CMD:
                self._log_to_ui("è®¾å¤‡æ— æ³•è¯†åˆ«å¼€å§‹å‘½ä»¤ï¼Œè¯·æ£€æŸ¥å›ºä»¶ç‰ˆæœ¬ã€‚")
                self._reset_detection_state()
            else:
                self._log_to_ui(f"è®¾å¤‡è¿”å›æœªçŸ¥çŠ¶æ€ç  {status_code}ï¼Œæ“ä½œå¤±è´¥ã€‚")
                self._reset_detection_state()

    def on_detection_timeout(self):
        self._log_to_ui("é”™è¯¯: å¥åº·ç›‘æµ‹è¶…æ—¶ (90ç§’)ï¼Œè¯·é‡è¯•ã€‚")
        self._reset_detection_state()

    def show_history_window(self):
        """æ˜¾ç¤ºå†å²æ•°æ®çª—å£"""
        # æ£€æŸ¥å®ä¾‹æ˜¯å¦å­˜åœ¨æˆ–å·²ä¸å¯è§ï¼Œé˜²æ­¢åˆ›å»ºå¤šä¸ªçª—å£
        if self.history_window_instance is None or not self.history_window_instance.isVisible():
            # å°† db_handler ä¸­çš„ db_file è·¯å¾„ä¼ é€’ç»™å†å²çª—å£
            self.history_window_instance = HistoryWindow(
                db_path=self.db_handler.db_file, 
                parent=self
            )
        self.history_window_instance.show()
        self.history_window_instance.activateWindow() # æ¿€æ´»çª—å£åˆ°å‰å°

    def on_health_data_received(self, data: bytes):
        if self.detection_timeout_timer.isActive():
            self.detection_timeout_timer.stop()

        self.startup_data_loaded = True
        self._stop_blinking()
        self._stop_countdown()
        self._log_to_ui(f"æ”¶åˆ°å¥åº·æ•°æ®: {data.hex(' ').upper()}")
            
        if len(data) >= 16:
            try:
                format_string = '<BBBBBBBBBBBBBI'
                unpacked_data = struct.unpack(format_string, data)
                
                health_metrics = dict(zip(self.metric_keys, unpacked_data))
                print(health_metrics)

                self._update_data_labels(health_metrics)
                self.db_handler.save_record_if_new(list(unpacked_data))
                
                # ä½“æ£€æˆåŠŸï¼Œé‡ç½®çŠ¶æ€
                self._reset_detection_state()

            except struct.error as e:
                self._log_to_ui(f"è§£æå¥åº·æ•°æ®å¤±è´¥: {e}")
        else:
            self._log_to_ui(f"è­¦å‘Š: å¥åº·æ•°æ® payload é•¿åº¦ä¸æ­£ç¡® (æ”¶åˆ° {len(data)} bytes)ã€‚")

    def _on_countdown_tick(self):
        if self.countdown_remaining > 0:
            self.countdown_remaining -= 1
        
        if self.start_button:
            if self.countdown_remaining > 0:
                self.start_button.setText(f"{self.countdown_remaining}ç§’")
            else:
                self.start_button.setText("å¤„ç†ä¸­...")

        if self.countdown_remaining <= 0 and self.countdown_timer.isActive():
            self.countdown_timer.stop()

    def _stop_countdown(self):
        if self.countdown_timer.isActive():
            self.countdown_timer.stop()
        self.countdown_remaining = 0

    def on_mouse_data_received(self, payload: bytes):
        """å¤„ç†æ”¶åˆ°çš„é¼ æ ‡ç´¯è®¡æ•°æ®ï¼Œæ›´æ–°ç•Œé¢å¹¶å†™å…¥æ•°æ®åº“ã€‚"""
        try:
            result = self.mouse_processor.process_payload(payload)
            self._log_to_ui(
                f"æ”¶åˆ°é¼ æ ‡æ•°æ®: è·ç¦»={result['distance_px']}px (~{result['distance_m_str']}), "
                f"L={result['left_click']}, M={result['mid_click']}, R={result['right_click']}"
            )
            self._update_mouse_labels(
                result['distance_px'],
                result['left_click'],
                result['mid_click'],
                result['right_click']
            )
            self.startup_mouse_loaded = True
        except Exception as e:
            self._log_to_ui(f"è§£æ/å¤„ç†é¼ æ ‡æ•°æ®å¤±è´¥: {e}")

    def _update_data_labels(self, data_dict: dict):
        for key, value in data_dict.items():
            if key in self.value_labels:
                display_value = value
                if key == 'cardiac':
                    # å°†å¿ƒè¾“å‡ºå€¼é™¤ä»¥10ï¼Œå¹¶æ ¼å¼åŒ–ä¸ºä¸€ä½å°æ•°çš„æµ®ç‚¹æ•°
                    try:
                        numeric_value = float(value)
                        display_value = f"{numeric_value / 10.0:.1f}"
                    except (ValueError, TypeError):
                        # å¦‚æœè½¬æ¢å¤±è´¥ï¼Œåˆ™æŒ‰åŸæ ·æ˜¾ç¤º
                        display_value = str(value)

                self.value_labels[key].setText(str(display_value))
        self._log_to_ui(f"ç•Œé¢æ•°æ®å·²æ›´æ–°ã€‚")

    def _update_mouse_labels(self, distance: int, left: int, mid: int, right: int):
        if self.label_distance:
            # ä½¿ç”¨ç±³åˆ¶å­—ç¬¦ä¸²å±•ç¤ºï¼›è‹¥å¤„ç†å™¨ä¸å¯ç”¨åˆ™å…œåº•ä¸ºåƒç´ å€¼
            try:
                meters_text = None
                if hasattr(self, 'mouse_processor') and self.mouse_processor:
                    meters_text = self.mouse_processor.pixels_to_meters_str(distance)
                self.label_distance.setText(meters_text if meters_text else str(distance))
            except Exception:
                self.label_distance.setText(str(distance))
            
        if self.label_leftclick:
            self.label_leftclick.setText(str(left))
        if self.label_midclick:
            self.label_midclick.setText(str(mid))
        if self.label_rightclick:
            self.label_rightclick.setText(str(right))
        self._log_to_ui("é¼ æ ‡æ•°æ®å·²æ›´æ–°åˆ°ç•Œé¢ã€‚")

    def _center_window(self):
        """å°†çª—å£ç§»åŠ¨åˆ°å±å¹•ä¸­å¿ƒ"""
        screen = QApplication.primaryScreen()
        if screen:
            screen_geometry = screen.availableGeometry()
            # æ‰‹åŠ¨è®¡ç®—å±…ä¸­ä½ç½®ï¼Œé¿å…ä¾èµ–æœªæ˜¾ç¤ºçš„ frameGeometry
            x = (screen_geometry.width() - self.width()) // 2
            y = (screen_geometry.height() - self.height()) // 2
            self.move(x, y)

    def _load_ui(self, ui_path: str) -> None:
        loader = QUiLoader()
        ui_file = QFile(ui_path)
        if not ui_file.exists():
            raise FileNotFoundError(f"UI æ–‡ä»¶æœªæ‰¾åˆ°: {ui_path}")
        if not ui_file.open(QFile.ReadOnly):
            raise IOError(f"æ— æ³•æ‰“å¼€ UI æ–‡ä»¶: {ui_path}")
        loaded = loader.load(ui_file)
        ui_file.close()
        if loaded is None:
            raise RuntimeError(f"åŠ è½½ UI å¤±è´¥: {ui_path}")

        # --- æå–æ ·å¼è¡¨å¹¶åº”ç”¨åˆ°å…¨å±€ï¼Œä»¥ä¾¿å­çª—å£ï¼ˆå¦‚å†å²çª—å£ï¼‰ä¹Ÿèƒ½ç»§æ‰¿æ ·å¼ ---
        app_style = loaded.styleSheet()
        if app_style:
            QApplication.instance().setStyleSheet(app_style)
            loaded.setStyleSheet("") # æ¸…é™¤æ§ä»¶è‡ªèº«çš„æ ·å¼è¡¨ï¼Œé¿å…åŒé‡åº”ç”¨

        if isinstance(loaded, QMainWindow):
            copied_title = loaded.windowTitle()
            copied_size = loaded.size()
            central = loaded.takeCentralWidget()
            self.setCentralWidget(central)
            self.setWindowTitle(copied_title)
            self.resize(copied_size) 
            loaded.deleteLater()
        else:
            self.setWindowTitle(loaded.windowTitle())
            self.resize(loaded.size())
            self.setCentralWidget(loaded)

    def _init_status_bar(self):
        self.statusBar().setStyleSheet("QStatusBar::item { border: none; }")
        self.status_icon = QLabel()
        self.status_label = QLabel("æœªè¿æ¥")
        self.statusBar().addWidget(self.status_icon)
        self.statusBar().addWidget(self.status_label)
        self._update_status_disconnected()

        # --- æ–°å¢ï¼šä¸º status_icon ç»‘å®šç‚¹å‡»äº‹ä»¶ ---
        self.status_icon.mouseReleaseEvent = self._on_status_icon_clicked

    def _on_status_icon_clicked(self, event):
        """å¤„ç†çŠ¶æ€å›¾æ ‡ç‚¹å‡»äº‹ä»¶"""
        if self.status_label.text() == "å·²è¿æ¥":
            self._log_to_ui("æ‰‹åŠ¨å‘é€è®¾å¤‡çŠ¶æ€æ£€æµ‹æŒ‡ä»¤...")
            self.serial_worker.send_frame(const.CMD_DEVICE_STATUS_CHECK)
        else:
            self._log_to_ui("è®¾å¤‡æœªè¿æ¥ï¼Œæ— æ³•å‘é€æŒ‡ä»¤ã€‚")

    def _update_status_connected(self):
        self.status_icon.setText("ğŸŸ¢")
        self.status_label.setText("å·²è¿æ¥")

    def _update_status_disconnected(self):
        self.status_icon.setText("ğŸ”´")
        self.status_label.setText("æœªè¿æ¥")

    def _create_emoji_icon(self, emoji_char, size=64):
        pixmap = QPixmap(size, size)
        pixmap.fill(Qt.GlobalColor.transparent)
        painter = QPainter(pixmap)
        
        font = QFont("Segoe UI Emoji", size - 10) 
        font.setPixelSize(size - 10)
        painter.setFont(font)
        
        font_metrics = QFontMetrics(font)
        text_width = font_metrics.horizontalAdvance(emoji_char)
        text_rect = font_metrics.boundingRect(emoji_char)
        y_pos = (pixmap.height() - text_rect.height()) / 2 + font_metrics.ascent()
        x_pos = (pixmap.width() - text_width) / 2

        painter.drawText(int(x_pos), int(y_pos), emoji_char)
        painter.end()
        return QIcon(pixmap)

    def _toggle_icon(self):
        if self.is_heart_icon:
            current_icon = self.icon_white_heart
        else:
            current_icon = self.icon_heart
        
        self.setWindowIcon(current_icon)
        self.tray_icon.setIcon(current_icon)
        self.is_heart_icon = not self.is_heart_icon

    def _handle_tray_activation(self, reason):
        if reason in (QSystemTrayIcon.ActivationReason.Trigger, QSystemTrayIcon.ActivationReason.DoubleClick):
            self.show_window()

    def _is_detection_in_progress(self) -> bool:
        # æ£€æµ‹ä¸­ï¼šæœ‰è¶…æ—¶å®šæ—¶å™¨æˆ–å€’è®¡æ—¶å®šæ—¶å™¨åœ¨è¿è¡Œ
        return (self.detection_timeout_timer.isActive() or
                (hasattr(self, 'countdown_timer') and self.countdown_timer.isActive()))

    def _reset_detection_state(self):
        """å°†ä¸å¥åº·æ£€æµ‹ç›¸å…³çš„UIå’Œè®¡æ—¶å™¨é‡ç½®åˆ°åˆå§‹çŠ¶æ€ã€‚"""
        self._stop_blinking()
        self._stop_countdown()
        if self.detection_timeout_timer.isActive():
            self.detection_timeout_timer.stop()
        if self.start_button:
            self.start_button.setEnabled(True)
            self.start_button.setText("å¼€å§‹ä½“æ£€")

    def _start_blinking(self):
        self._log_to_ui("å¼€å§‹é—ªçƒ...")
        if not self.blink_timer.isActive():
            self.blink_timer.start()

    def _stop_blinking(self):
        if self.blink_timer.isActive():
            self.blink_timer.stop()
            self.setWindowIcon(self.icon_heart)
            self.tray_icon.setIcon(self.icon_heart)
            self.is_heart_icon = False
            self._log_to_ui("åœæ­¢é—ªçƒã€‚")

    def show_window(self):
        # æ¢å¤çª—å£æ—¶ï¼Œè‹¥ä»åœ¨æ£€æµ‹ä¸­åˆ™ä¿æŒé—ªçƒ
        if not self._is_detection_in_progress():
            self._stop_blinking()
        self.show()
        self.activateWindow()

    def hide_window(self):
        self.hide()

    def exit_app(self):
        self.tray_icon.hide()
        self._shutdown_cleanup()
        QApplication.quit()

    def closeEvent(self, event):
        # ç‚¹å‡»çª—å£å³ä¸Šè§’å…³é—­æŒ‰é’®æ—¶ï¼Œæœ€å°åŒ–åˆ°æ‰˜ç›˜ï¼Œä¸é€€å‡ºåº”ç”¨ã€ä¸æ–­å¼€ä¸²å£
        self._log_to_ui("å·²æœ€å°åŒ–åˆ°æ‰˜ç›˜ã€‚é€šè¿‡æ‰˜ç›˜å›¾æ ‡å¯å†æ¬¡æ‰“å¼€ï¼Œæˆ–é€‰æ‹©é€€å‡ºã€‚")
        event.ignore()
        self.hide_window()

    def _shutdown_cleanup(self):
        """é€€å‡ºåº”ç”¨å‰çš„èµ„æºæ¸…ç†ï¼šæ–­å¼€ä¸²å£å¹¶åœæ­¢çº¿ç¨‹ã€‚"""
        try:
            self.serial_worker.disconnect_serial()
        except Exception:
            pass
        try:
            if self.serial_thread.isRunning():
                self.serial_thread.quit()
                if not self.serial_thread.wait(1000):
                    self._log_to_ui("è­¦å‘Š: ä¸²å£çº¿ç¨‹æœªèƒ½æ­£å¸¸åœæ­¢ã€‚")
        except Exception:
            pass
